DS-3531: Nice numeric cuts

chrisfacer opened the pull request:
Functions to help users automatically group numeric data based on different methods. Factors whose labels represent ranges (e.g. age, income) are managed by a second function specifically for merging ranges.

Commit:
Adding NiceNumericCuts and MergeRangeCategories

Adding NiceNumericCuts and MergeRangeCategories for DS-3531. Added test cases and testthat files.

Commit:
Improvements to numeric cuts and range merging

- Ensure that when numeric cuts is applied to a factor or character than any non-numeric levels are filled back in rather than setting them as missing data.
- Ensure that the specified number of categories to merge always applies only to the numeric portion of the data. Before, if the user asked for 5 categories and there was 1 non-numeric category, the result would contain 4 numeric categories plus the non-numeric, but now it will contain 5 numeric categories plus the non-numeric.

Commit:
Protect against empty strings

Protect against empty strings

Commit:
Better handling of empty start/end values

Better handling of empty start/end values

Commit:
Fix to last update

Fix to last update

Commit:
Better messages [revdep skip]

Better messages [revdep skip]

Commit:
Typo in label message [revdep skip]

Typo in label message [revdep skip]

Commit:
Fix again [revdep skip]

Fix again [revdep skip]

Commit:
Typo [revdep skip]

Typo [revdep skip]

Commit:
Better warnings and default num.categories [revdep skip]

Better warnings and default num.categories [revdep skip]

Commit:
Fix issue with bad variable labels [revdep skip]

Fix issue with bad variable labels [revdep skip]

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+
+#' Create factor new factor variables from existing factors whose
```
Typo, "factor" is repeated.

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+#' Create factor new factor variables from existing factors whose
+#' labels contain numeric ranges, like 18 to 24 or $2,000 to $3,000,
+# 'by merging adjacent ranges.
+#'
```
The roxygen should start with the title for the help page, followed by a blank line, then a longer description about what the function does. E.g.
```
#' This is the title, a short description of the function
#'
#' This is a longer description of the function
#' @param x Numeric vector
.
.
.
```
If you look at your Rd files, they have a title section, but no description section.

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+#'      numeric ranges when the number is not explicitly included in the 
+#'      level labels. For example, when the lowest level is "Under 18", you
+#'      may wish to specify that the lowest possible value in this level
+#'      is 12, 15, 0, etc. This helps when you wish to 
```
Incomplete sentence.

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+#'      level labels. For example, when the lowest level is "Under 18", you
+#'      may wish to specify that the lowest possible value in this level
+#'      is 12, 15, 0, etc. This helps when you wish to 
+#' @param upper.bound As above, for the higher end of the range.
```
Would be good to describe what the default value `NULL` does.

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+#' @param grouping.mark A character indicating how thousands are grouped within
+#'      the labels of the input data. Usally a comma, but sometimes a period
+#'      for example in European locales.
+#' @param decimals.mark A character indicating how decimal values are separated
```
Should state that this is/can be a regular expression.

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+    # Compute counts. Needs to be updated to handle weights?
+    counts = table(raw.data)
+
+    # Try to pull appart labels into numbers and non-numbers
```
Typo

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+                                upper.bound = NULL) {
+
+    if (is.data.frame(input.data)) {
+        raw.data = unlist(input.data)
```
Our style guide (and all our packages) use `<-` for assignment, not `=` https://docs.displayr.com/wiki/Flip_Project_R_Style_Guide#Assignment

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+        decimals.mark = escapeRegex(decimals.mark)
+
+        x = gsub(grouping.mark, "", x)
+        x = gsub(decimals.mark, "\\.", x)
```
"\\." can be "." since the second argument is not a regex. Am actually surprised it works with "\\.", `gsub` seems to just ignore the `\\`.

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+        x = gsub(grouping.mark, "", x)
+        x = gsub(decimals.mark, "\\.", x)
+
+        numbers = as.numeric(unlist(regmatches(x, gregexpr("(?>-)*[[:digit:]]+\\.*[[:digit:]]*", x, perl=TRUE))))
```
Can you add a comment please explaining this regex pattern? Not sure why the `*` on `\\.` and `(?>-)`. Should it be `?` for 0 or 1 matches (i.e. optional). Also thought that `(?>-)` would be `(?>=-)` or `(?>!-)` for positive or negative lookbehind.

chrisfacer replied:
Nice catch. Teach me to copy from StackExchange.

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+
+        delimiter.candidates = non.numbers[-c(1, length(non.numbers))]
+
+        all.deliiters.same = all(delimiter.candidates == delimiter.candidates[1])
```
Not used

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+                                          values = values, 
+                                          target.number = num.categories)
+
+    for (j in 1L:length(merge.solution)) {
```
A bit more efficient to do `seq_along(merge.solution)`

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+    if (!is.ranges) {
+        stop("Could not detect sufficient numeric data to merge ranges. When combining ",
+            "data whose labels represent ranges, all labels should contain pairs of ",
+            "numbers except for those labels at the higher and lower end of the range.",
```
Need space after "range."

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+        stop("Could not detect sufficient numeric data to merge ranges. When combining ",
+            "data whose labels represent ranges, all labels should contain pairs of ",
+            "numbers except for those labels at the higher and lower end of the range.",
+            "In this case there are ", length(which(singles)), "labels which do not ",
```
Needs a space before "labels"

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+        values = merge.data$range
+
+    } else {
+        stop("Method, '", method[1],"' not recognized. Should be ",
```
The "fancy" R way to do this is say `method <- match.arg(method)` at the start of the function. This will automatically throw an error if method doesn't match one of the options in the function signature and remove the need to specify the `[1]` when referring to method throughout the function.

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+#' @param input.data A \code{data.frame} or \code{vector} containing factors
+#'      to be merged.
+#' @param num.categories An integer indicating the number of categories
+#'      in the new factors.
```
... excluding categories that don't contain ranges.

mwmclean commented on R/nicenumericcuts.R:
```diff
@@ -0,0 +1,657 @@
+#' @param num.categories An integer sepcifying the number of categories for
+#'      the new factor. This does not apply for \code{percentiles} or for
+#'      \code{Custom} because in those cases the number of categories is
+#'      imlpied by the \code{percents} and \code{custom.breaks} argumens
```
Two typos

mwmclean commented on R/nicenumericcuts.R:
```diff
@@ -0,0 +1,657 @@
+#'
+#' @importFrom flipU ConvertCommaSeparatedStringToVector
+#' @importFrom plyr mapvalues
+#' @importFrom stats var quantile
```
You may want to add `#' @return` to document the returned object and `#' @examples`  to provide example usage.

mwmclean commented on R/nicenumericcuts.R:
```diff
@@ -0,0 +1,657 @@
+                      type = quantile.type)
+        
+        if (length(which(duplicated(qq))) > 0) {
+            warning("Some percentiles are empty in the range you have specified and will not be shown");
```
Missing "."

mwmclean commented on R/nicenumericcuts.R:
```diff
@@ -0,0 +1,657 @@
+                      na.rm = TRUE, include.lowest = TRUE, 
+                      type = quantile.type)
+        
+        if (length(which(duplicated(qq))) > 0) {
```
More efficient: `anyDuplicated(qq)`

mwmclean commented on R/nicenumericcuts.R:
```diff
@@ -0,0 +1,657 @@
+
+    
+
+    # factors = vapply(classes, FUN = function (x) return("factor" %in% x), FUN.VALUE = logical(1))
```
Any reason to leave this commented out code in?

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+        }
+
+        if (length(numbers) + 1 != length(non.numbers)) {
+            stop("Chris, you've got the splitting wrong!")
```
Change this to something user-friendly just in case?

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+    if (is.unsorted(label.data$second.value, na.rm = TRUE)) {
+        stop("The numbers in the data labels suggest that the ranges are inconsistent or overlapping. ",
+            "This can happen if labels refer to different units of measurement, for example some ",
+            "referring to months and others referring to years. Please esnure your labels contain consistent units.")
```
ensure

mwmclean commented on R/mergecategoriescontainingranges.R:
```diff
@@ -0,0 +1,456 @@
+    lower.boundary.open.phrases = c("less than", "under", "fewer than", "lower than", "below", "<", "smaller than")
+    lower.boundary.closed.phrases = c( "up to", "<=", "and less", "and under", "and lower", "and below", "and smaller")
+
+    upper.boundary.open.phrases = c("greater than", "over", "more than", "above", ">")
```
Could add "larger than" and perhaps "older than" for ages and time/date ranges.

mwmclean reviewed (commented):
Can we support examples like `ilock$Q4` where there is extra text before/after the numeric range?

Commit:
Changes from Math's review [revdep skip]

Changes from Math's review [revdep skip]. Includes:
- Fixed typos
- Improved regex
- Improved argument checking for 'method'

Commit:
Improved options for equal.width and custom [revdep skip]

Improved options for equal.width and custom [revdep skip]
- Add option to custom to allow the user to force the min and max values to be included in the cuts regardless of whether these two values are included in custom.breaks.
- Add option to equal.width to allow the user to specify the width of the interval, which overrides their choice of num.categories.

mwmclean committed:
Fix typo in test name [ci skip]

mwmclean commented on R/nicenumericcuts.R:
```diff
@@ -0,0 +1,679 @@
+                           closed.bottom.string = " and below",
+                           open.top.string = "More than ",
+                           closed.top.string = " and over",
+                           equal.intervals.start = 0,
```
```suggestion
                           equal.intervals.start = min(input.data),
```

chrisfacer replied:
setting NULL default instead as data may not be numeric until coercion.

mwmclean commented on R/nicenumericcuts.R:
```diff
@@ -0,0 +1,679 @@
+                           open.top.string = "More than ",
+                           closed.top.string = " and over",
+                           equal.intervals.start = 0,
+                           equal.intervals.end = 100,
```
```suggestion
                           equal.intervals.end = max(input.data),
```

chrisfacer replied:
setting NULL default instead as data may not be numeric until coercion.

Commit:
Better defaults for equal.intervals [revdep skip]

Better defaults for equal.intervals [revdep skip]
